src/snowflake/snowpark/_internal/debug_utils.py

```diff
@@ -0,0 +1,200 @@
+#
+# Copyright (c) 2012-2025 Snowflake Computing Inc. All rights reserved.
+#
+
+from functools import cached_property
+import os
+import sys
+from typing import Dict, List, Optional
+import itertools
+
+from snowflake.snowpark._internal.ast.batch import get_dependent_bind_ids
+from snowflake.snowpark._internal.ast.utils import __STRING_INTERNING_MAP__
+import snowflake.snowpark._internal.proto.generated.ast_pb2 as proto
+
+UNKNOWN_FILE = "__UNKNOWN_FILE__"
+SNOWPARK_PYTHON_DATAFRAME_TRANSFORM_TRACE_LENGTH = (
+    "SNOWPARK_PYTHON_DATAFRAME_TRANSFORM_TRACE_LENGTH"
+)
+
+
+class DataFrameTraceNode:
+    """A node representing a dataframe operation in the DAG that represents the lineage of a DataFrame."""
+
+    def __init__(self, batch_id: int, stmt_cache: Dict[int, proto.Stmt]) -> None:
+        self.batch_id = batch_id
+        self.stmt_cache = stmt_cache
+
+    @cached_property
+    def children(self) -> set[int]:
+        """Returns the batch_ids of the children of this node."""
+        return get_dependent_bind_ids(self.stmt_cache[self.batch_id])
```

Comment:
This helper method will return all dependencies of a given statement (anything that requires an AST ID including UDxFs and Sprocs), not just the direct dependencies. I see later you're expecting `node.children` to just be the direct dependencies to get the lineage.